lib/ui/compositing/scene.cc

```diff
@@ -102,8 +98,42 @@ Dart_Handle Scene::toImage(uint32_t width,
   return Picture::RasterizeToImage(picture, width, height, raw_image_callback);
 }
 
+void Scene::RasterizeToImageSync(uint32_t width,
+                                 uint32_t height,
+                                 Dart_Handle raw_image_handle) {
+  auto* dart_state = UIDartState::Current();
+  auto unref_queue = dart_state->GetSkiaUnrefQueue();
+  auto snapshot_delegate = dart_state->GetSnapshotDelegate();
+  auto raster_task_runner = dart_state->GetTaskRunners().GetRasterTaskRunner();
+
+  auto image = CanvasImage::Create();
+  auto dl_image = DlDeferredImageGPU::Make(SkISize::Make(width, height));
+  image->set_image(dl_image);
+
+  fml::TaskRunner::RunNowOrPostTask(
+      raster_task_runner,
+      [snapshot_delegate, unref_queue, dl_image = std::move(dl_image),
+       layer_tree = layer_tree_, width, height]() {
+        auto display_list =
+            layer_tree->Flatten(SkRect::MakeWH(width, height),
+                                snapshot_delegate.get()->GetTextureRegistry());
+
+        sk_sp<SkImage> sk_image;
+        std::string error;
+        std::tie(sk_image, error) = snapshot_delegate->MakeGpuImage(
+            display_list, dl_image->dimensions());
+        if (sk_image) {
+          dl_image->set_image(std::move(sk_image));
+        } else {
+          dl_image->set_error(std::move(error));
+        }
+      });
+
+  image->AssociateWithDartWrapper(raw_image_handle);
+}
+
 std::unique_ptr<flutter::LayerTree> Scene::takeLayerTree() {
-  return std::move(layer_tree_);
+  return nullptr;
```

Comment:
Changing to shared_ptr almost works, except that everything that uses this API expects a unique_ptr.

Maybe I can hold onto the parts used to construct the LayerTree and then build either a shared_ptr or unique_ptr on the fly?

The root layer used to construct the layer tree is itself a shared_ptr...

Reply:
its a lot, let me dig in some more.

Reply:
I updated the PR with all the changes needed to compile this